File: mwt-legacy/internal/cli/add.go
package cli

import (
	"fmt"
	"path/filepath"

	"github.com/spf13/cobra"
)

type addOptions struct {
	repos      []string
	from       string
	noSetup    bool
	continueOn bool
}

func newAddCmd(d Deps) *cobra.Command {
	var opts addOptions
	cmd := &cobra.Command{
		Use:   "add <branch>",
		Short: "Create worktrees for the branch across configured repos",
		Long: `Create a git worktree for <branch> in each selected repo (serial).

For each repo: mkdir -p the worktree parent, git worktree add, then run
configured setup steps (unless --no-setup). Use --from to create the branch
from a start point when it does not exist yet.

With --continue, remaining repos still run after a failure; the command
exits non-zero if any repo failed.`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runAdd(d, args[0], opts)
		},
	}
	cmd.Flags().StringSliceVar(&opts.repos, "repos", nil, "subset of config repos (comma-separated or repeated)")
	cmd.Flags().StringVar(&opts.from, "from", "", "start point when creating a missing branch (git worktree add -b)")
	cmd.Flags().BoolVar(&opts.noSetup, "no-setup", false, "skip setup steps after creating the worktree")
	cmd.Flags().BoolVar(&opts.continueOn, "continue", false, "best-effort: continue after a repo failure; still exit non-zero if any failed")
	return cmd
}

func runAdd(d Deps, branch string, opts addOptions) error {
	if branch == "" {
		return fmt.Errorf("branch is required")
	}
	cfg, err := d.loadConfig()
	if err != nil {
		return err
	}
	repos, err := selectRepos(cfg, opts.repos)
	if err != nil {
		return err
	}
	if len(repos) == 0 {
		return fmt.Errorf("no repos selected")
	}

	out := d.stdout()
	repoWidth := maxRepoWidth(repos)
	return runSerial(repos, opts.continueOn, func(repo string) error {
		ctx, err := resolveRepo(cfg, repo, branch)
		if err != nil {
			return err
		}
		parent := filepath.Dir(ctx.WorktreePath)
		if err := d.mkdirAll(parent, 0o755); err != nil {
			return fmt.Errorf("mkdir %s: %w", parent, err)
		}
		if err := d.git().Add(ctx.MainPath, ctx.WorktreePath, branch, opts.from); err != nil {
			return err
		}
		if !opts.noSetup {
			if err := runSetupForRepo(d.setup(), cfg, ctx); err != nil {
				return err
			}
		}
		fmt.Fprintf(out, "%-*s  %s\n", repoWidth, repo, displayPath(cfg.MetaRoot, ctx.WorktreePath))
		return nil
	})
}

File: src/cli/add.rs
//! `mwt add <branch>` — create worktrees for the branch across configured repos.

use std::io::Write;

use clap::Args;

use crate::cli::deps::{Deps, SetupRunner};
use crate::cli::shared::{display_path, max_repo_width, resolve_repo, run_serial, select_repos};

#[derive(Debug, Args)]
pub struct AddOptions {
    /// Subset of config repos (comma-separated or repeated).
    #[arg(long, value_delimiter = ',')]
    pub repos: Vec<String>,

    /// Start point when creating a missing branch (`git worktree add -b`).
    #[arg(long)]
    pub from: String,

    /// Skip setup steps after creating the worktree.
    #[arg(long)]
    pub no_setup: bool,

    /// Best-effort: continue after a repo failure; still exit non-zero if any failed.
    #[arg(long)]
    pub r#continue: bool,
}

pub fn run(
    deps: &Deps,
    branch: &str,
    opts: &AddOptions,
    setup_runner: &mut dyn SetupRunner,
    out: &mut dyn Write,
) -> Result<(), String> {
    if branch.is_empty() {
        return Err("branch is required".to_string());
    }
    let cfg = (deps.load_config)().map_err(|e| format!("load config: {e}"))?;
    let repos = select_repos(&cfg, &opts.repos)?;
    if repos.is_empty() {
        return Err("no repos selected".to_string());
    }

    let setup: &[crate::config::SetupStep] = &cfg.setup;
    let repo_width = max_repo_width(&repos);

    run_serial(&repos, opts.r#continue, |repo| {
        let ctx = resolve_repo(&cfg, repo, branch)?;
        let parent = ctx.worktree_path.parent().ok_or_else(|| {
            format!(
                "worktree path has no parent: {}",
                ctx.worktree_path.display()
            )
        })?;
        (deps.mkdir_all)(parent)
            .map_err(|e| format!("mkdir {parent}: {e}", parent = parent.display()))?;
        deps.git
            .add(&ctx.main_path, &ctx.worktree_path, branch, &opts.from)?;
        if !opts.no_setup {
            setup_runner
                .run(&ctx, setup)
                .map_err(|e| format!("setup: {e}"))?;
        }
        writeln!(
            out,
            "{repo:<repo_width$}  {path}",
            path = display_path(&cfg.meta_root, &ctx.worktree_path.to_string_lossy())
        )
        .map_err(|e| e.to_string())?;
        Ok(())
    })
}
